### PR TITLE
[do not review] Cmake transitive header inclusion

### PR DESCRIPTION
**Pull request by krocard**
Enhance our cmake configuration by:
 - avoiding cach configuration duplication
 - lib now declare their public header
 - lib no longer need to explicitly include the include dir of used library

**FIXME**: Require cmake 3.0 to work. Consider replacing PUBLIC/PRIVATE by LINK_PUBLIC/LINK_PRIVATE in target_link_libraries to support cmake 2.8.
See: http://www.cmake.org/cmake/help/v2.8.12/cmake.html#command:target_link_libraries

<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39141931%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39141983%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39142309%22%2C%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39142414%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%2010a52c5841620ec900da50a2b83595a63135d4fe%20test/catch/CMakeLists.txt%2037%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39141983%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22TODO%3A%20Raise%20an%20error%20with%20this%20message.%22%2C%20%22created_at%22%3A%20%222015-09-10T09%3A46%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20test/catch/CMakeLists.txt%3AL1-42%22%7D%2C%20%22Pull%2010a52c5841620ec900da50a2b83595a63135d4fe%20xmlserializer/CMakeLists.txt%2016%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39142309%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22...%20part%20of%20libxml2%27s%20INTERFACE_COMPILE_DEFINITIONS%5Cr%5CnIf%20it%20is%20remove%20this%20line.%22%2C%20%22created_at%22%3A%20%222015-09-10T09%3A50%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20xmlserializer/CMakeLists.txt%3AL53-69%22%7D%2C%20%22Pull%2010a52c5841620ec900da50a2b83595a63135d4fe%20xmlserializer/CMakeLists.txt%208%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39142414%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Check%20if%20%24%7BLIBXML2_INCLUDE_DIR%7D%20is%20not%20part%20of%20libxml2%27s%20INTERFACE_INCLUDE_DIRECTORIES.%5Cr%5CnIf%20it%20is%2C%20remove%20this%20line.%22%2C%20%22created_at%22%3A%20%222015-09-10T09%3A51%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20xmlserializer/CMakeLists.txt%3AL53-69%22%7D%2C%20%22Pull%2010a52c5841620ec900da50a2b83595a63135d4fe%20parameter/CMakeLists.txt%2020%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/01org/parameter-framework/pull/200%23discussion_r39141931%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%2APUBLIC%2A%22%2C%20%22created_at%22%3A%20%222015-09-10T09%3A45%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6862950%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/krocard%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20parameter/CMakeLists.txt%3AL103-120%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 10a52c5841620ec900da50a2b83595a63135d4fe parameter/CMakeLists.txt 20'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/200#discussion_r39141931'>File: parameter/CMakeLists.txt:L103-120</a></b>
- <a href='https://github.com/krocard'></a> *PUBLIC*
- [ ] <a href='#crh-comment-Pull 10a52c5841620ec900da50a2b83595a63135d4fe test/catch/CMakeLists.txt 37'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/200#discussion_r39141983'>File: test/catch/CMakeLists.txt:L1-42</a></b>
- <a href='https://github.com/krocard'></a> TODO: Raise an error with this message.
- [ ] <a href='#crh-comment-Pull 10a52c5841620ec900da50a2b83595a63135d4fe xmlserializer/CMakeLists.txt 16'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/200#discussion_r39142309'>File: xmlserializer/CMakeLists.txt:L53-69</a></b>
- <a href='https://github.com/krocard'></a> ... part of libxml2's INTERFACE_COMPILE_DEFINITIONS
If it is remove this line.
- [ ] <a href='#crh-comment-Pull 10a52c5841620ec900da50a2b83595a63135d4fe xmlserializer/CMakeLists.txt 8'></a> &nbsp;<b><a href='https://github.com/01org/parameter-framework/pull/200#discussion_r39142414'>File: xmlserializer/CMakeLists.txt:L53-69</a></b>
- <a href='https://github.com/krocard'></a> Check if ${LIBXML2_INCLUDE_DIR} is not part of libxml2's INTERFACE_INCLUDE_DIRECTORIES.
If it is, remove this line.

<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/200?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/01org/parameter-framework/pull/200?approve=1'></a>&nbsp;<a href='https://github.com/01org/parameter-framework/pull/200'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dawagner** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/01org/parameter-framework/pull/200'>Refresh</a>.<a href='#crh-update-none'></a>

File: parameter/CMakeLists.txt [hunk 1]
```diff
@@ -103,15 +103,17 @@ add_library(parameter SHARED
...
 find_library(dl dl)
-target_link_libraries(parameter xmlserializer remote-processor pfw_utility dl)
+target_link_libraries(parameter 
+    # unfortunatly xmlSink and xmlSource need to be exposed to the plugins
+    PUBLIC xmlserializer
+    PRIVATE remote-processor pfw_utility dl)
+
+target_include_directories(parameter
+    PUBLIC include log/include
+    # FIXE: define . as an include directory only for plugins
```
Comment (krocard) [inline parameter/CMakeLists.txt]: *PUBLIC*

File: test/catch/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+if(BUILD_TESTING)
+    # Add catch unit test framework
+    # TODO Use gtest as it is the team recommendation
+    #      Unfortunately gtest is very hard to setup as not binary distributed
+    #      catch is only one header so it is very easy
+    # Catch can be downloaded from:
+    # https://raw.github.com/philsquared/Catch/master/single_include/catch.hpp
+    # Then append the download folder to the CMAKE_INCLUDE_PATH variable or
+    # copy it in a standard location (/usr/include on most linux distribution).
```
Comment (krocard) [inline test/catch/CMakeLists.txt]: TODO: Raise an error with this message.

File: xmlserializer/CMakeLists.txt [hunk 3]
```diff
@@ -53,11 +53,16 @@ if(NOT XML2_XINCLUDE_SUPPORT)
...
+target_include_directories(xmlserializer
+    PRIVATE ${LIBXML2_INCLUDE_DIR}
+    PUBLIC .)
 
-target_link_libraries(xmlserializer ${LIBXML2_LIBRARIES})
+target_link_libraries(xmlserializer
+    PUBLIC pfw_utility # For NonCopyable
+    PRIVATE ${LIBXML2_LIBRARIES})
+
+# TODO: check if the libxml2 definitions are not already part of the interface
```
Comment (krocard) [inline xmlserializer/CMakeLists.txt]: ... part of libxml2's INTERFACE_COMPILE_DEFINITIONS
If it is remove this line.

File: xmlserializer/CMakeLists.txt [hunk 4]
```diff
@@ -53,11 +53,16 @@ if(NOT XML2_XINCLUDE_SUPPORT)
             "please install a version of libxml2 supporting it.")
 endif()
 
-include_directories(
-    ${LIBXML2_INCLUDE_DIR}
-    "${PROJECT_SOURCE_DIR}/utility")
+target_include_directories(xmlserializer
+    PRIVATE ${LIBXML2_INCLUDE_DIR}
```
Comment (krocard) [inline xmlserializer/CMakeLists.txt]: Check if ${LIBXML2_INCLUDE_DIR} is not part of libxml2's INTERFACE_INCLUDE_DIRECTORIES.
If it is, remove this line.

**Comment by krocard** [comment]
@krocard Please resubmit against `master` branch.

**Comment by krocard** [comment]
Resubmited on master. Removing the `resubmit` tag.

File: parameter/CMakeLists.txt (hunk 1 above)
Comment (krocard) [inline parameter/CMakeLists.txt]: :+1:

File: test/catch/CMakeLists.txt (hunk 2 above)
Comment (krocard) [inline test/catch/CMakeLists.txt]: :+1: see #273

File: xmlserializer/CMakeLists.txt (hunk 3 above)
Comment (krocard) [inline xmlserializer/CMakeLists.txt]: It is not :+1:

File: xmlserializer/CMakeLists.txt (hunk 4 above)
Comment (krocard) [inline xmlserializer/CMakeLists.txt]: It is not :+1:

**Pull request closed without merging**